torch/distributed/_tensor/dispatch.py

```diff
@@ -237,18 +241,19 @@ def default_tensor(spec: DTensorSpec) -> torch.Tensor:
         # for random ops, set RNG offset
         assert isinstance(mesh, DeviceMesh)
         if op_call in random_ops and is_rng_supported_mesh(mesh):
-            dtensor_arg = arg_list[0]
-            old_offset = _get_rng_offset(mesh)
-            set_pre_op_offset(dtensor_arg._spec)
-
-        # run local op computation with potentially modified args/kwargs
-        local_tensor_args = cast(Tuple[object, ...], local_tensor_args)
-        local_tensor_kwargs = cast(Dict[str, object], local_tensor_kwargs)
-        local_results = op_call(*local_tensor_args, **local_tensor_kwargs)
-
-        # if op is a random op, adjust Philox RNG state to maintain synchronization
-        if op_call in random_ops and is_rng_supported_mesh(mesh):
-            set_post_op_offset(dtensor_arg._spec, old_offset)
+            # TODO: use the new RNG primitive???
+            devices = [torch.cuda.current_device()]
+            with torch.random.fork_rng(devices):
+                dtensor_arg = arg_list[0]
+                # set up offset before op call
+                seed, offset = get_rng_state(mesh)
+                _set_pre_op_rng_state(dtensor_arg._spec, seed, offset)
```

Comment:
first getting RNG, set pre_op_rng, after computation, we set_post_rng, this looks a lot like a context manager, I think we should encapsulate the set_pre_rng and post_rng stuff in a context manager instead to make the code clean.